engine/lib/analyses/flatten_ast.ml

```diff
@@ -0,0 +1,49 @@
+open Base
+open Utils
+
+module%inlined_contents Make (F : Features.T) = struct
+  module FA = F
+  module A = Ast.Make (F)
+  open Ast
+
+  let rec flatten_ast (x : A.expr) : A.expr list =
```

Comment:
Here, `flatten_ast` can be implemented directly using [visitors](http://gallium.inria.fr/~fpottier/visitors/manual.pdf), which are derived for the entire AST:
```ml
let rec flatten_ast : A.expr -> A.expr list =
    (object
       inherit [_] A.expr_reduce as super
       inherit [_] list_monoid as m
       method visit_t _ _ = m#zero
       method visit_mutability (_f : unit -> _ -> _) () _ = m#zero
       method visit_expr f e = e :: super#visit_expr f e
    end)#visit_expr ()
```

given some `list_monoid` like:
```ml
class ['s] list_monoid =
    object
      inherit ['s] VisitorsRuntime.monoid
      method private zero = []
      method private plus = List.append
    end
```

The nice thing with this is that the code will be robust against AST changes.